File: leetcode/cpp/tree_utils.cpp
#include <iostream>
#include <vector>
#include <queue>

#include "tree_utils.h"

using namespace std;

void BFS(TreeNode *root) {
    queue<TreeNode*> q;
    q.push(root);

    while(!q.empty()) {
        TreeNode *p = q.front();
        q.pop();

        if(!p) {
            cout << "N" << " ";
            continue;
        }

        cout << p->val << " ";

        q.push(p->left);
        q.push(p->right);
    }

    cout << endl;
}

void inorder_rec(TreeNode *root) {
    if(!root)
        return;

    inorder_rec(root->left);
    cout << root->val << " ";
    inorder_rec(root->right);
}

void inorder(TreeNode *root) {
    cout << "inorder = { ";
    inorder_rec(root);
    cout << "}" << endl;
}

void preorder_rec(TreeNode *root) {
    if(!root)
        return;

    cout << root->val << " ";
    preorder_rec(root->left);
    preorder_rec(root->right);
}

void preorder(TreeNode *root) {
    cout << "preorder = { ";
    preorder_rec(root);
    cout << "}" << endl;
}

void postorder_rec(TreeNode *root) {
    if(!root)
        return;

    postorder_rec(root->left);
    postorder_rec(root->right);
    cout << root->val << " ";
}

void postorder(TreeNode *root) {
    cout << "postorder = { ";
    postorder_rec(root);
    cout << "}" << endl;
}

/*
 * https://stackoverflow.com/questions/36802354/print-binary-tree-in-a-pretty-way-using-c
 */
void printBT(const std::string& prefix, const TreeNode* node, bool isLeft) {
    if( node != nullptr ) {
        cout << prefix;

        cout << (isLeft ? "├──" : "└──" );

        // print the value of the node
        cout << node->val << endl;

        // enter the next tree level - left and right branch
        printBT( prefix + (isLeft ? "│   " : "    "), node->left, true);
        printBT( prefix + (isLeft ? "│   " : "    "), node->right, false);
    }
}

void printBT(const TreeNode* node) {
    printBT("", node, false);
}

TreeNode* build_tree(vector<string> &arr) {
    if(arr.size() == 0) {
        return nullptr;
    }
    queue<TreeNode*> q;
    TreeNode *root = new TreeNode(stoi(arr[0]));
    q.push(root);

    for(int i=1; i < arr.size(); i+=2) {
        TreeNode *p = q.front();
        q.pop();

        p->left  = (arr[i]  == "N") ? nullptr : new TreeNode(stoi(arr[i]));

        if(i < arr.size()-1) {
            p->right = (arr[i+1] == "N") ? nullptr : new TreeNode(stoi(arr[i+1]));
        }

        if(p->left)
            q.push(p->left);

        if(p->right)
            q.push(p->right);
    }

    return root;
}

File: leetcode/python3/tree_utils.py


from typing import List
from queue import Queue

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def BuildTree(arr: List[str]) -> TreeNode:
    q = Queue()

    v = int(arr[0])
    root = TreeNode(val=v)

    q.put(root)

    for i in range(1, len(arr), 2):
        node: TreeNode = q.get()

        if arr[i] != "null":
            v = int(arr[i])
            node.left = TreeNode(val=v)
            q.put(node.left)

        if i + 1 < len(arr) and arr[i + 1] != "null":
            v = int(arr[i+1])
            node.right = TreeNode(val=v)
            q.put(node.right)

    return root


def _printTreeUtil(prefix: str, root: TreeNode, level: int, isleft: bool):
    if root is None:
        return

    s: str = ""

    if level != 0:
        if isleft:
            s = prefix + "├──"
            prefix += "│   "
        else:
            s = prefix + "└──"
            prefix += "    "

    print("{}{}".format(s, root.val))

    if root.left is None and root.right is None:
        return

    _printTreeUtil(prefix, root.left, level + 1, True)
    _printTreeUtil(prefix, root.right, level+1, False)


def PrintTree(root: TreeNode):
    _printTreeUtil("", root, 0, False)


if __name__ == "__main__":
    arr = ["-10", "9", "20", "null", "null", "15", "7"]
    root = BuildTree(arr)
    PrintTree(root)
